### PR TITLE
[Clean up] Simplify UpsertConfig

### PR DESCRIPTION
**Pull request by deemoliu**
Instructions:
#cleanup

As we have more and more properties key in the upsertConfig, simplified UpsertConfig to avoid using the constructor with all properties as parameters.

This can cause compilation backward incompatible every time a new property is added. 
This PR cleans up the usage of UpsertConfig constructor with all properties, and replace with empty constructor and setters.

### CODE REVIEW

**Comment by deemoliu** [comment]
cc: @chenboat @Jackie-Jiang @yupeng9

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java [hunk 1]
```diff
@@ -77,6 +77,10 @@ public UpsertConfig(@JsonProperty(value = "mode", required = true) Mode mode,
     _hashFunction = hashFunction == null ? HashFunction.NONE : hashFunction;
   }
 
+  public UpsertConfig(Mode mode) {
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:80]: Let's use an empty constructor which is required for the json deserialization

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java [hunk 2]
```diff
@@ -44,16 +44,16 @@ public enum Strategy {
...
 
   @JsonPropertyDescription("default upsert strategy for partial mode")
-  private final Strategy _defaultPartialUpsertStrategy;
+  private Strategy _defaultPartialUpsertStrategy;
 
   @JsonPropertyDescription("Column for upsert comparison, default to time column")
-  private final String _comparisonColumn;
+  private String _comparisonColumn;
 
   @JsonCreator
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:58]: Remove this annotation and add `@Deprecated`

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java (hunk 1 above)
Comment (deemoliu) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java]: gotcha, thanks!

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java (hunk 2 above)
Comment (deemoliu) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java]: Thanks, addressed the comment.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java [hunk 3]
```diff
@@ -97,4 +99,24 @@ public Strategy getDefaultPartialUpsertStrategy() {
...
+
+  public void setMode(Mode mode) {
+    _mode = mode;
+  }
+
+  public void setHashFunction(HashFunction hashFunction) {
+    _hashFunction = hashFunction;
+  }
+
+  public void setPartialUpsertStrategies(Map<String, Strategy> partialUpsertStrategies) {
```
Comment (KKcorps) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:111]: Can we add java doc here telling that key in this map should be columnName.

Comment (deemoliu) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:111]: Gotcha, let me add it.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java [hunk 4]
```diff
@@ -41,21 +40,21 @@ public enum Strategy {
   }
 
   @JsonPropertyDescription("Upsert mode.")
-  private final Mode _mode;
+  private Mode _mode;
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:43]: Consider making `_mode` default to `FULL` and `_hashFunction` default to `NONE`, which is the most common setup and these 2 fields cannot be `null`

Comment (deemoliu) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:43]: Thanks @Jackie-Jiang 
For `_mode` the value can be `NONE` for non-upsert tables. 
Is there any side effect on changing the default value to "FULL"?

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java [hunk 5]
```diff
@@ -97,4 +101,40 @@ public Strategy getDefaultPartialUpsertStrategy() {
...
+  public void setMode(Mode mode) {
+    _mode = mode;
+  }
+
+  public void setHashFunction(HashFunction hashFunction) {
+    _hashFunction = hashFunction;
+  }
+
+  /**
+   * PartialUpsertStrategies maintains the mapping of merge strategies per columns.
```
Comment (chenboat) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java:114]: columns --> column

Comment (deemoliu) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/UpsertConfig.java]: thanks, nice catch! addressed comment.

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**